### PR TITLE
Add lint for 'field_reassign_with_default` #568

### LINKED ISSUE #568: suggest struct constructor instead of mutable object + field mod
If you have a struct where all fields are public, then you should never do `let mut a = A::default(); a.x = 42;` but instead use a constructor with `.. A::default()` at the end:

``` rust
#[derive(Default)]
struct A {
    i: i32,
    j: i64,
}

fn main() {
    let mut a: A = Default::default();
    a.i = 42;
}

fn should_be() {
    let a = A {
        i: 42,
        .. Default::default()
    };
}
```

### PR DESCRIPTION
**Pull request by hegza**
changelog: Add lint for field_reassign_with_default that checks if mutable object + field modification is used to edit a binding initialized with Default::default() instead of struct constructor.

Fixes #568

Notes:
- Checks for reassignment of one or more fields of a binding initialized with Default::default().
- Implemented using EarlyLintPass, might be future proofed better with LateLintPass.
- Does not trigger if Default::default() is used via another type implementing Default.
- This is a re-open of [PR#4761](https://github.com/rust-lang/rust-clippy/pull/4761), but I couldn't figure out how to re-open that one so here's a new one with the requested changes :S

### CODE REVIEW

**Comment by Manishearth** [comment]
r? @ebroto 

might take some time to get to this

**Comment by ebroto** [comment]
Just a heads-up, I'm half-through the review, should have the time to finish it later.

**Review by ebroto — CHANGES_REQUESTED** [review]
Thanks for your pull request! I think this lint can be useful, especially for those not familiar with functional update syntax.

IMO this lint is trickier than it seems... I think the main issue I see with the current iteration is that as we don't recurse into subexpressions, with code like this:
```rust
let mut a: A = Default::default();
a.i = 42;
if true {
    a.j = 21;
}
```
we would:
1) Not catch the assignment to `j`
2) Worse, suggest making the binding immutable, suggesting code that does not compile.

To make this work, we would probably need a visitor that looks for assignments to the fields of the struct. But what if those assignments are inside a loop? What about function calls that take `&mut` to one of the fields? We should avoid linting in those cases (see [here](https://github.com/rust-lang/rust-clippy/blob/6220dff504b1ac1a949e34562db47fd058378f5d/clippy_lints/src/loops.rs#L1781) for some examples of visitors).

I'm inclined to keep things simple and avoid recursing into subexpressions like we do now, but avoid suggesting an immutable binding. That way if there are other assignments we don't suggest code that does not compile, and if there are none rustc's `unused_mut` will trigger after applying the suggestion. 

What do y'all think @flip1995 @Manishearth ?

**Comment by hegza** [comment]
Thanks for finding the edge cases. I tried to reason about it originally and thought that that first one wasn't possible but my mental model was incorrect. There's a lot but I think I should be able to do another try with late lint pass and the offered suggestions. I'll get it through eventually since it's useful.

**Comment by ebroto** [comment]
Ping from triage @hegza. Did you have the time to address the review comments? If you are stuck on something we can try to help :)

**Comment by hegza** [comment]
@ebroto On it. I made a pass on the trivial issues already and signed them off as resolved.

It then took a while to assimilate the rest of what you said into a plan but I feel like I'm progressing and I'm at least currently having a good plan on what to tackle and in which order. I'll get in contact with you if I get stuck.

In detail:
- lint on duplicate assignment to a field should be easy by asserting uniqueness of each element,
- not suggesting mutable is also easy,
- writing my test case on not adding Default::default() to suggestion should clarify how to proceed with that one for me,
- writing my test case on tuples should clarify that one also,
- also test case on needless_update will clarify,
- auto-apply is also easy after the other things are done.

File: clippy_lints/src/field_reassign_with_default.rs
Comment (hegza) [reply clippy_lints/src/field_reassign_with_default.rs]: "If field is initialized with `Default::default()...`, do you mean like below?

```
let mut a: A = Default::default();
a.i = Default::default();
```

File: clippy_lints/src/field_reassign_with_default.rs
Comment (hegza) [reply clippy_lints/src/field_reassign_with_default.rs:98]: Tuples are now ignored.

**Comment by hegza** [comment]
@ebroto now the only things missing are [fields initialized as Defaul::default()](https://github.com/rust-lang/rust-clippy/pull/5911#discussion_r473394511) and auto-apply. The auto-apply seems complicated because I'm not familiar with the concept of FRU. Can you clarify?

Everything else should be done, but I'll leave it here until I can get a comment on the above two issues.

File: clippy_lints/src/field_reassign_with_default.rs
Comment (hegza) [reply clippy_lints/src/field_reassign_with_default.rs]: Category set as style.

File: clippy_lints/src/field_reassign_with_default.rs
Comment (hegza) [reply clippy_lints/src/field_reassign_with_default.rs]: Known problems now only contains a note about assignments to patterns of tuple-types.

File: clippy_lints/src/field_reassign_with_default.rs
Comment (hegza) [reply clippy_lints/src/field_reassign_with_default.rs]: Done.

File: clippy_lints/src/field_reassign_with_default.rs
Comment (hegza) [reply clippy_lints/src/field_reassign_with_default.rs]: Transformed to LatePass as suggested.

File: clippy_lints/src/field_reassign_with_default.rs
Comment (hegza) [reply clippy_lints/src/field_reassign_with_default.rs]: Used a HashMap to keep track of assignments, then did a comparison to check if all fields are set and changed output based on that.

File: clippy_lints/src/field_reassign_with_default.rs
Comment (hegza) [reply clippy_lints/src/field_reassign_with_default.rs]: Fixed and test added.

File: clippy_lints/src/field_reassign_with_default.rs
Comment (ebroto) [inline clippy_lints/src/field_reassign_with_default.rs:6]:
```suggestion
use rustc_middle::ty::{self, Adt, Ty};
```

File: clippy_lints/src/field_reassign_with_default.rs
Comment (ebroto) [inline clippy_lints/src/field_reassign_with_default.rs:47]:
```suggestion
        let binding_statements_using_default =
```
Although maybe you added the annotation for clarity?

File: clippy_lints/src/field_reassign_with_default.rs
Comment (ebroto) [inline clippy_lints/src/field_reassign_with_default.rs:97]:
```suggestion
                    let ext_with_default = !fields_of_type(binding_type)
```

File: clippy_lints/src/field_reassign_with_default.rs
Comment (ebroto) [inline clippy_lints/src/field_reassign_with_default.rs:117]:
```suggestion
                        first_assign.unwrap().span,
```

File: clippy_lints/src/field_reassign_with_default.rs
Comment (ebroto) [inline clippy_lints/src/field_reassign_with_default.rs:133]:
```suggestion
fn enumerate_bindings_using_default<'tcx>(cx: &LateContext<'tcx>, block: &Block<'_>) -> Vec<(usize, Symbol, Ty<'tcx>)> {
```

File: clippy_lints/src/field_reassign_with_default.rs
Comment (ebroto) [inline clippy_lints/src/field_reassign_with_default.rs:143]:
```suggestion
                if let PatKind::Binding(_, _, ident, _) = local.pat.kind;
                // that are not tuples
                let ty = cx.typeck_results().pat_ty(local.pat);
                if !matches!(ty.kind(), ty::Tuple(_));
```
We can reuse the `if_chain` for this check

File: clippy_lints/src/field_reassign_with_default.rs
Comment (ebroto) [inline clippy_lints/src/field_reassign_with_default.rs:157]:
```suggestion
```

File: clippy_lints/src/field_reassign_with_default.rs
Comment (ebroto) [inline clippy_lints/src/field_reassign_with_default.rs:178]:
```suggestion
fn field_reassigned_by_stmt<'tcx>(this: &Stmt<'tcx>, binding_name: Symbol) -> Option<(Ident, &'tcx Expr<'tcx>)> {
```
For consistency with most of the codebase

File: clippy_lints/src/field_reassign_with_default.rs
Comment (ebroto) [inline clippy_lints/src/field_reassign_with_default.rs:199]:
```suggestion
fn fields_of_type(ty: Ty<'_>) -> Vec<Ident> {
```
This can be simplified

File: clippy_lints/src/field_reassign_with_default.rs
Comment (ebroto) [inline clippy_lints/src/field_reassign_with_default.rs:203]:
```suggestion
            let variant = &adt.non_enum_variant();
```

File: tests/ui/field_reassign_with_default.rs
Comment (ebroto) [inline tests/ui/field_reassign_with_default.rs:16]: Something I missed in my first review (my apologies): this would currently be linted by `default_trait_access`. Since that lint is pedantic, this one should get preference.

I suggest putting both lints in a `default_trait` module, and having `default_trait_access` check if `field_reassign_with_default` would trigger, and avoid linting in that case.

File: tests/ui/field_reassign_with_default.rs
Comment (ebroto) [inline tests/ui/field_reassign_with_default.rs:75]: We should avoid linting if one assignment refers to another field, e.g.
```rust
    let mut x = A::default();
    x.i = 42;
    x.j = 21 + x.i as i64;
```
Given that `j: 21 + x.i as i64` would not compile

File: clippy_lints/src/field_reassign_with_default.rs
Comment (ebroto) [inline clippy_lints/src/field_reassign_with_default.rs:105]: The order of the fields in the suggestion should match the order of the replaced statements. 

Consider this case:
```rust
    let mut y = A::default();
    y.j = order_matters() as i64;
    y.i = order_matters();
```
With a function like this:
```rust
fn order_matters() -> i32 {
    static mut VAL: i32 = 0;
    unsafe {
        VAL += 1;
        VAL
    }
}
```
The suggestion would change the values that `i` and `j` get.

File: clippy_lints/src/field_reassign_with_default.rs
Comment (ebroto) [inline clippy_lints/src/field_reassign_with_default.rs:120]: We could add something like `and remove the relevant assignments` to the suggestion to make it clearer

**Review by ebroto — CHANGES_REQUESTED** [review]
Thanks for taking the time to apply the suggestions!

Most of my new comments are related to style, but some of them require rethinking the lint a bit. As I said in the last review, I think this lint is trickier than it seems.

File: clippy_lints/src/field_reassign_with_default.rs
Comment (ebroto) [reply clippy_lints/src/field_reassign_with_default.rs]: Yes, exactly.

File: clippy_lints/src/field_reassign_with_default.rs
Comment (ebroto) [reply clippy_lints/src/field_reassign_with_default.rs]: Given that this lint is requiring a considerable effort, we may want to leave the auto-fixes for a followup issue/PR.

**Comment by ebroto** [comment]
@hegza by FRU I mean "Functional Record Update", so the same thing we are trying to recommend in the suggestion. About the automatic fixes, see my new comment, we can leave them for a follow-up.

File: tests/ui/field_reassign_with_default.rs
Comment (hegza) [reply tests/ui/field_reassign_with_default.rs:16]: How would you check if a lint would trigger?

Comment (ebroto) [reply tests/ui/field_reassign_with_default.rs:16]: Usually having two lints in the same module and targeting the same kind of node would make this easier, but in this case even though both lints target expressions, `field_reassign_with_default` needs more context so it works at the block level.

We could maybe leverage this: knowing that this lint will find the relevant expression before `default_trait_access`, we could record the linted span and have `default_trait_access` check if the expression it would lint lies in the same span, bailing out in that case.

Comment (hegza) [reply tests/ui/field_reassign_with_default.rs:16]: Do you mean that it needs to be implemented via some sort of a manual mechanism? Eg. LintB borrowing a feature from LintA to check if LintA would trigger, before doing its own work.

Comment (ebroto) [reply tests/ui/field_reassign_with_default.rs:16]: Yes, there is no mecanism to disable a lint temporarily AFAIK.

In this case I mean to keep track of the spans linted by `field_reassign_with_default` in a collection, and have `default_trait_access` check that collection before linting. This should work given that `check_block` will run before the `check_expr` that would trigger `default_trait_access`.

File: tests/ui/field_reassign_with_default.rs
Comment (hegza) [reply tests/ui/field_reassign_with_default.rs:75]: Fixed. We now cancel linting for the binding `x` if a reference to it's symbol is encountered on the right-hand side of field assignment.

File: clippy_lints/src/field_reassign_with_default.rs
Comment (hegza) [reply clippy_lints/src/field_reassign_with_default.rs:105]: Order is now retained.

**Comment by flip1995** [comment]
Ping from triage @ebroto: I think there was some progress since your last review, but I'm not so sure. Can you take a look?

@hegza If you're waiting for another round of review, don't hesitate to ping the reviewer.

**Comment by ebroto** [comment]
Yeah I saw it but was waiting for the remaining remark to be solved (conflicting suggestions with `default_trait_access`) as there will probably be a good amount of code moved.

@hegza would you like me to review your current changes before that?

**Review by ebroto — CHANGES_REQUESTED** [review]
Two things I found while reviewing the current state of the PR.

Besides that, I think this is close to being ready. Thanks for pushing this forward!

**Comment by hegza** [comment]
> 
> 
> Ping from triage @ebroto: I think there was some progress since your last review, but I'm not so sure. Can you take a look?
> 
> @hegza If you're waiting for another round of review, don't hesitate to ping the reviewer.

There's progress (albeit slow; other responsibilities cluttering my weekends etc. the usual) but I'm still in the middle of some of the previous changes. I'll ping once I've decided on the next step for each review and recommendation.

The main problem I'm working on is combining the lint with `default_trait_access`. I think it's easily doable but I'm slightly(!) scared of the refactoring work.

Thanks for another round of review, anyway :D

**Comment by ebroto** [comment]
>There's progress (albeit slow; other responsibilities cluttering my weekends etc. the usual) but I'm still in the middle of some of the previous changes. I'll ping once I've decided on the next step for each review and recommendation.

Of course, no rush, this can wait :)

>The main problem I'm working on is combining the lint with default_trait_access. I think it's easily doable but I'm slightly(!) scared of the refactoring work.

Don't hesitate to ping us regarding the refactoring in case you find a blocker.

Again, thanks for sticking with this. I would be very happy to see this merged given the effort you've already put into it.

File: tests/ui/field_reassign_with_default.rs
Comment (hegza) [reply tests/ui/field_reassign_with_default.rs:16]: @ebroto ok I thought about it for a while, and **I can't figure out a way to talk from one lint to another.** The purpose is to keep track of spans linted by field_reassign_with_default for default_trait_access to read.

The reason for why I can't get the lints to talk is that the two lints are both implemented for two separate unit-structs (= no data), and the implementations are called from lib.rs. Static globals are also out-of-the-question I would assume. Caching into LateContext seems ugly too. Total merge of the two lints does not solve the communication issue since the backing data structure is still a unit-struct.

It's the last thing missing.

Comment (ebroto) [reply tests/ui/field_reassign_with_default.rs:16]: Sorry for not being more precise! 

`declare_lint_pass!` is not the only way to create a new pass, you can do that "manually" and then use `impl_lint_pass!`. This way you can add state to the pass, which can be used by both lints.

I've added the changes myself to a new `default` module. I will dig into the last round of review ASAP.

**Review by ebroto — CHANGES_REQUESTED** [review]
The only thing missing are the remarks from the last round of review: https://github.com/rust-lang/rust-clippy/pull/5911#pullrequestreview-500405037

After that I think we can merge this! :)

File: clippy_lints/src/field_reassign_with_default.rs
Comment (hegza) [reply clippy_lints/src/field_reassign_with_default.rs]: This is currently covered. I verified it with a test case.

**Comment by hegza** [comment]
@ebroto I think the cases you refer to are already fixed but I hadn't signed them off. I wrote another test-case to verify that as well.

I think we're done (maybe?) If you find more, let me know and I'll get back to it.

**Review by ebroto — APPROVED** [review]
>I think the cases you refer to are already fixed 

Indeed, we were just missing the test you added.

---

LGTM! Could you squash your commits into one? After that we can merge this!

**Comment by hegza** [comment]
Squashed @ebroto.

**Review by ebroto — APPROVED** [review]

**Comment by ebroto** [comment]
@bors r+

Thanks for sticking with this during this really long review process :heart:

**Pull request merged**
